### ISSUE TITLE
Training mode and testing mode

### ISSUE THREAD

**Issue by findmyway** [JuliaReinforcementLearning/ReinforcementLearning.jl#179]
Today I was asked *what is the intended way to evaluate agents after training?* Then I noticed I had never written down the design decisions anywhere. In the next release, I'll move things here into docs.

In **supervised learning**, a dataset is usually split into *training*, *testing* (and also *verification*) parts. We train a model based on the training dataset and then compare different hyper-parameters (or even different models) on the test dataset.  Sometimes, the behavior of the same model between training and testing datasets are different. A typical one is the dropout layer. In [Flux.jl](https://github.com/FluxML/Flux.jl/blob/1bc9ac1d49c08a894d309882f8edb644bb8ee0c3/docs/src/models/layers.md#testmode), it's handled like this:

> Many normalisation layers behave differently under training and inference (testing). By default, Flux will automatically determine when a layer evaluation is part of training or inference. Still, depending on your use case, it may be helpful to manually specify when these layers should be treated as being trained or not. For this, Flux provides `Flux.testmode!`. When called on a model (e.g. a layer or chain of layers), this function will place the model into the mode specified.

```
Flux.testmode!
trainmode!
```

So in previous versions of `ReinforcementLearning.jl`, we just reused the semantic of these two functions and applied them to `Agent` and any `AbstractPolicy` (**in-place** updating). However, things in RL are slightly different. Two important concepts in RL are **exploration** and **exploitation**. In two different modes, we may change the degree of **exploration** and **exploitation**. Simply setting `training` or `testing` mode is not enough.

In the latest master branch, instead of updating policies in-place in different modes, we decide to create a new policy instance and dispatch different behaviors based on the type.

- In `run(policy, env, stop_condition, hook)`, we don't assume whether we're running in the training mode or testing mode. Only dispatch the `policy` to different stages:
    - `policy(stage, env)`. By default we do nothing. But if `policy` is an `Agent`, then we apply `update!(agent.policy, agent.trajectory, env, stage)`

Now let's answer the original question:

 *what is the intended way to evaluate agents after training?* 

Example:

```julia
training_hook = DoEveryNEpisode() do t, agent, env
    policy = agent.policy
    // set exploration or the testing mode of inner DNN model
    run(policy, env, test_stop_condition, test_hook)
end

run(agent, env, training_stop_condition, training_hook) 
```

**Comment by Sid-Bhatia-0** [comment]
Thank you for the explanation @findmyway 

Would it be possible to provide an end-to-end example of training an agent and then testing it (everything end-to-end in a single file)? Perhaps something in the format of RLZoo experiments (like [this](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/experiments/rl_envs/JuliaRL_BasicDQN_CartPole.jl) one)? Even a simple end-to-end example like this would be super helpful!

**Comment by findmyway** [comment]
I think experiments under the atari folder in RLZoo follow the pipeline I mentioned above.

**Comment by Sid-Bhatia-0** [comment]
Thanks. I'll check it out.

**Comment by findmyway** [comment]
This has been discussed in https://juliareinforcementlearning.org/blog/an_introduction_to_reinforcement_learning_jl_design_implementations_thoughts/#21_the_general_workflow
